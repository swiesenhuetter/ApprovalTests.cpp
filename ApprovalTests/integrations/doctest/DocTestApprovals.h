#pragma once

#include "ApprovalTests/namers/ApprovalTestNamer.h"
#include "ApprovalTests/integrations/CheckFileMacroIsAbsolute.h"

#if defined(APPROVALS_DOCTEST_EXISTING_MAIN)
#define APPROVALS_DOCTEST
#define DOCTEST_CONFIG_IMPLEMENT
#elif defined(APPROVALS_DOCTEST)
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#endif

#ifdef APPROVALS_DOCTEST

// begin-snippet: required_header_for_doctest
#include <doctest/doctest.h>
// end-snippet

namespace ApprovalTests
{
    // anonymous namespace to prevent compiler -Wsubobject-linkage warnings
    // This is OK as this code is only compiled on main()
    namespace
    {
        struct AbstractReporter : doctest::IReporter
        {
            virtual void report_query(const doctest::QueryData&) override
            {
            }
            // called when the whole test run starts
            virtual void test_run_start() override
            {
            }

            // called when the whole test run ends (caching a pointer to the input doesn't make sense here)
            virtual void test_run_end(const doctest::TestRunStats&) override
            {
            }

            // called when a test case is started (safe to cache a pointer to the input)
            virtual void test_case_start(const doctest::TestCaseData&) override
            {
            }

#if 20305 <= DOCTEST_VERSION
            // called when a test case is reentered because of unfinished subcases (safe to cache a pointer to the input)
            virtual void test_case_reenter(const doctest::TestCaseData&) override
            {
            }
#endif

            // called when a test case has ended
            virtual void test_case_end(const doctest::CurrentTestCaseStats&) override
            {
            }

            // called when an exception is thrown from the test case (or it crashes)
            virtual void test_case_exception(const doctest::TestCaseException&) override
            {
            }

            // called whenever a subcase is entered (don't cache pointers to the input)
            virtual void subcase_start(const doctest::SubcaseSignature&) override
            {
            }

            // called whenever a subcase is exited (don't cache pointers to the input)
            virtual void subcase_end() override
            {
            }

            // called for each assert (don't cache pointers to the input)
            virtual void log_assert(const doctest::AssertData&) override
            {
            }

            // called for each message (don't cache pointers to the input)
            virtual void log_message(const doctest::MessageData&) override
            {
            }

            // called when a test case is skipped either because it doesn't pass the filters, has a skip decorator
            // or isn't in the execution range (between first and last) (safe to cache a pointer to the input)
            virtual void test_case_skipped(const doctest::TestCaseData&) override
            {
            }
        };

        struct DocTestApprovalListener : AbstractReporter
        {
            TestName currentTest;

            // constructor has to accept the ContextOptions by ref as a single argument
            explicit DocTestApprovalListener(const doctest::ContextOptions& /*in*/)
            {
            }

            void test_case_start(const doctest::TestCaseData& testInfo) override
            {

                currentTest.sections.emplace_back(testInfo.m_name);
                currentTest.setFileName(testInfo.m_file);
                ApprovalTestNamer::currentTest(&currentTest);
            }

            void test_case_end(const doctest::CurrentTestCaseStats& /*in*/) override
            {

                while (!currentTest.sections.empty())
                {
                    currentTest.sections.pop_back();
                }
            }

            void subcase_start(const doctest::SubcaseSignature& signature) override
            {

#if DOCTEST_VERSION >= 20307
                currentTest.sections.emplace_back(signature.m_name.c_str());
#else
                currentTest.sections.emplace_back(signature.m_name);
#endif
            }

            void subcase_end() override
            {

                currentTest.sections.pop_back();
            }
        };
    }
}

REGISTER_LISTENER("approvals", 0, ApprovalTests::DocTestApprovalListener);

#endif // APPROVALS_DOCTEST
