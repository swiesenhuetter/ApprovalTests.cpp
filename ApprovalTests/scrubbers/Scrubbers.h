#pragma once

#include <string>
#include <functional>
#include <iostream>
#include <regex>
#include <ApprovalTests/utilities/StringUtils.h>

namespace ApprovalTests
{
    using Scrubber = std::function<std::string(const std::string&)>;
    namespace Scrubbers
    {
        inline std::string doNothing(const std::string& input)
        {
            return input;
        }

        inline std::string scrubGuid(const std::string& input)
        {
            static const std::regex regex("[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-["
                                          "0-9a-fA-F]{4}-[0-9a-fA-F]{12}");

            int matchNumber = 1;
            auto result = input;
            std::smatch m;
            while (std::regex_search(result, m, regex))
            {
                auto guid_match = m.str();
                auto replacement = "guid_" + std::to_string(matchNumber);
                result = StringUtils::replaceAll(result, guid_match, replacement);
                matchNumber += 1;
            }
            return result;
        }
    }
}
