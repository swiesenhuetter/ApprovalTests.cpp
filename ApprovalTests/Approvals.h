#pragma once

#include <string>
#include <functional>
#include <exception>

#include "ApprovalTests/ApprovalTestsVersion.h"
#include "ApprovalTests/core/FileApprover.h"
#include "reporters/DefaultReporter.h"
#include "reporters/DefaultReporterDisposer.h"
#include "ApprovalTests/core/Reporter.h"
#include "namers/ApprovalTestNamer.h"
#include "writers/ExistingFile.h"
#include "namers/ExistingFileNamer.h"
#include "namers/SubdirectoryDisposer.h"
#include "namers/DefaultNamerDisposer.h"
#include "scrubbers/Scrubbers.h"

namespace ApprovalTests
{

    class Approvals
    {
    private:
        Approvals() = default;

        ~Approvals() = default;

    public:
        static std::shared_ptr<ApprovalNamer> getDefaultNamer()
        {
            return DefaultNamerFactory::getDefaultNamer()();
        }

        static void verify(std::string contents,
                           const Reporter& reporter = DefaultReporter())
        {
            verifyScrubbed(contents, Scrubbers::doNothing, reporter);
        }

        static void verifyScrubbed(std::string contents,
                                   Scrubber scrubber,
                                   const Reporter& reporter = DefaultReporter())
        {
            verifyWithExtension(scrubber(contents), ".txt", reporter);
        }

        static void verifyWithExtension(std::string contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            StringWriter writer(contents, fileExtensionWithDot);
            FileApprover::verify(*getDefaultNamer(), writer, reporter);
        }

        static void verify(const ApprovalWriter& writer,
                           const Reporter& reporter = DefaultReporter())
        {
            FileApprover::verify(*getDefaultNamer(), writer, reporter);
        }

        template <typename T>
        using IsNotDerivedFromWriter =
            typename std::enable_if<!std::is_base_of<ApprovalWriter, T>::value,
                                    int>::type;

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verify(const T& contents,
                           const Reporter& reporter = DefaultReporter())
        {
            verify(StringUtils::toString(contents), reporter);
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verifyWithExtension(const T& contents,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            verifyWithExtension(
                StringUtils::toString(contents), fileExtensionWithDot, reporter);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verify(const T& contents,
                           Function converter,
                           const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            converter(contents, s);
            verify(s.str(), reporter);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotDerivedFromReporter<Function>>
        static void verifyWithExtension(const T& contents,
                                        Function converter,
                                        const std::string& fileExtensionWithDot,
                                        const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            converter(contents, s);
            verifyWithExtension(s.str(), fileExtensionWithDot, reporter);
        }

        static void verifyExceptionMessage(std::function<void(void)> functionThatThrows,
                                           const Reporter& reporter = DefaultReporter())
        {
            std::string message = "*** no exception thrown ***";
            try
            {
                functionThatThrows();
            }
            catch (const std::exception& e)
            {
                message = e.what();
            }
            verify(message, reporter);
        }

        template <typename Iterator>
        static void verifyAll(
            std::string header,
            const Iterator& start,
            const Iterator& finish,
            std::function<void(typename Iterator::value_type, std::ostream&)> converter,
            const Reporter& reporter = DefaultReporter())
        {
            std::stringstream s;
            if (!header.empty())
            {
                s << header << "\n\n\n";
            }
            for (auto it = start; it != finish; ++it)
            {
                converter(*it, s);
                s << '\n';
            }
            verify(s.str(), reporter);
        }

        template <typename Container>
        static void verifyAll(
            std::string header,
            const Container& list,
            std::function<void(typename Container::value_type, std::ostream&)> converter,
            const Reporter& reporter = DefaultReporter())
        {
            verifyAll<typename Container::const_iterator>(
                header, list.begin(), list.end(), converter, reporter);
        }

        template <typename T>
        static void verifyAll(std::string header,
                              const std::vector<T>& list,
                              const Reporter& reporter = DefaultReporter())
        {
            int i = 0;
            verifyAll<std::vector<T>>(
                header,
                list,
                [&](T e, std::ostream& s) { s << "[" << i++ << "] = " << e; },
                reporter);
        }

        template <typename T>
        static void verifyAll(const std::vector<T>& list,
                              const Reporter& reporter = DefaultReporter())
        {
            verifyAll<T>("", list, reporter);
        }

        static void verifyExistingFile(const std::string filePath,
                                       const Reporter& reporter = DefaultReporter())
        {
            ExistingFile writer(filePath);
            ExistingFileNamer namer(filePath);
            FileApprover::verify(namer, writer, reporter);
        }

        static SubdirectoryDisposer
        useApprovalsSubdirectory(std::string subdirectory = "approval_tests")
        {
            return SubdirectoryDisposer(subdirectory);
        }

        static DefaultReporterDisposer
        useAsDefaultReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return DefaultReporterDisposer(reporter);
        }

        static FrontLoadedReporterDisposer
        useAsFrontLoadedReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return FrontLoadedReporterDisposer(reporter);
        }

        static DefaultNamerDisposer useAsDefaultNamer(NamerCreator namerCreator)
        {
            return DefaultNamerDisposer(namerCreator);
        }
    };
}
