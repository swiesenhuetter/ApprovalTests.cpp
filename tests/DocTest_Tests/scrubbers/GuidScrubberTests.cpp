#include "doctest/doctest.h"
#include "ApprovalTests/scrubbers/Scrubbers.h"

// <iostream> is needed to fix linker error on XCode Release builds:
#include <iostream>
#include <ApprovalTests/Approvals.h>

using namespace ApprovalTests;

TEST_CASE("Input with single GUID")
{
    auto input = "2fd78d4a-ad49-447d-96a8-deda585a9aa5 and normal text";
    auto output = Scrubbers::scrubGuid(input);
    REQUIRE(output == "guid_1 and normal text");
}

TEST_CASE("Input with no GUIDs")
{
    auto input = "";
    auto output = Scrubbers::scrubGuid(input);
    REQUIRE(output == "");
}

TEST_CASE("Input with non-GUID")
{
    std::string input = "hello world";
    auto output = Scrubbers::scrubGuid(input);
    REQUIRE(output == input);
}

TEST_CASE("Input with non-GUID")
{

    std::string input = R"(
{
    child: {
        id: b34b4da8-090e-49d8-bd35-7e79f633a2ea
        parent1: 2fd78d4a-ad49-447d-96a8-deda585a9aa5
        parent2: 05f77de3-3790-4d45-b045-def96c9cd371
    }
    person: {
        name: mom
        id: 2fd78d4a-ad49-447d-96a8-deda585a9aa5
    }
    person: {
        name: dad
        id: 05f77de3-3790-4d45-b045-def96c9cd371
    }
}
)";
    auto output = Scrubbers::scrubGuid(input);
    Approvals::verifyScrubbed(input, Scrubbers::scrubGuid);
}
