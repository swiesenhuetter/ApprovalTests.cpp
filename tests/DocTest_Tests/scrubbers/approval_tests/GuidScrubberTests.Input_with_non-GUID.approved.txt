
{
    child: {
        id: guid_1
        parent1: guid_2
        parent2: guid_3
    }
    person: {
        name: mom
        id: guid_2
    }
    person: {
        name: dad
        id: guid_3
    }
}

